runtime/lua/vim/lsp.lua

```diff
@@ -669,6 +669,13 @@ function lsp.start_client(config)
         if client_ids[client_id] then
           client._on_attach(bufnr)
         end
+        -- local allow_incremental_sync = if_nil(client.config.flags.allow_incremental_sync, false)
+        -- if allow_incremental_sync then
+        if not client.cached_buffers then
+          client.cached_buffers = {}
+        end
+        client.cached_buffers[bufnr] = nvim_buf_get_lines(bufnr, 0, -1, true)
```

Comment:
```suggestion
        if not client.cached_buffers then
          client._cached_buffers = {}
        end
        client._cached_buffers[bufnr] = nvim_buf_get_lines(bufnr, 0, -1, true)
```

We probably want the cached buffers to be an implementation detail and hidden from users - or at least prefix with `_` to indicate that it's considered private.

Reply:
agree. later on (0.6) I hope to enable reading the old line contents from the undo header in the change callback, so double buffering won't be needed.

Reply:
Re: performance, I don't know. It seems fine on smaller files (not sure if you noticed differently). I think when using start/end_line to filter, it's just finding the difference (on average) over probably 160 characters (unless people are using incredibly long lines), so I think it's likely fine? Debouncing would be very tricky.

Reply:
> performance, I don't know. It seems fine on smaller files (not sure if you noticed differently). I think when using start/end_line to filter, it's just finding the difference (on average) over probably 160 characters (unless people are using incredibly long lines), so I think it's likely fine?

Yes I think you're right. If it's only computing the diff on the lines affected it should be fast enough.

Reply:
The only reason I didn't do this yet is I wanted a 1-to-1 comparison to vim-lsc so I could use it to debug differences in our compute_differences function, which I thought I had completely solved 😆